preprocess/split.py

```diff
@@ -1,20 +1,92 @@
 from abc import abstractmethod
+from ukb_loader import UKBDataLoader
+from numpy.random import seed, choice
+import pandas as pd
+pd.options.mode.chained_assignment = None # Shush
 
+import sys
+sys.path.append('../config')
+
+from config.path import data_root, valid_ids_path, ukb_loader_dir, ukb_pfile_path
+from config.split_config import split_map
+from utils.plink import run_plink
 
 class SplitBase(object):
-    def __init__(self, split_config):
+    def __init__(self, split_config: dict):
         self.split_config = split_config
 
     @abstractmethod
     def split(self):
         pass
+    
+    def split_by_ethnicity(self) -> pd.DataFrame:
+        loader = UKBDataLoader(ukb_loader_dir, 'split', '21000', ['31'])
+        df = pd.concat((loader.load_train(), loader.load_val(), loader.load_test()))
+        df.columns = ['sex', 'ethnic_background']
 
+        # Include FID/IID fields for plink to play nice with the output files
+        df['FID'] = df.index
+        df['IID'] = df.index
+        # Leave only those samples that passed population QC
+        pop_qc_ids = pd.read_csv(valid_ids_path, index_col='IID')
+        df = df.loc[df.index.intersection(pop_qc_ids.index)]
 
-class SplitIID(SplitBase):
-    def split(self):
-        pass
-
+        df.ethnic_background[df.ethnic_background.isna()] = 0.0
+        df.ethnic_background = df.ethnic_background.astype('int')
+        # Drop samples with missing/prefer not to answer ethnic background
+        df = df[~df.ethnic_background.isin([-1, -3, 0])]
+        # Map ethnic backgrounds to our defined splits
+        df['split_code'] = df.ethnic_background.map(split_map)
+        
+        return df
+    
+    def make_split_pgen(self, split_id_path: str, prefix: str) -> None:
+        run_plink(args_dict={'--pfile': ukb_pfile_path,
+                             '--out': prefix,
+                             '--keep': split_id_path},
+                  args_list=['--make-pgen'])
 
+    
+class SplitIID(SplitBase):
+    def split(self, make_pgen=True):
+        df = self.split_by_ethnicity()
```

Comment:
why do we need to run this function here? why not simply take df = df.loc[df['ethnic_background'] == 1001]?

Reply:
Agreed, tidying it up